client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java

```diff
@@ -2002,6 +2014,164 @@ protected static boolean convertExistsResponse(Response response) {
         return response.getStatusLine().getStatusCode() == 200;
     }
 
+    private Cancellable performClientRequestAsync(Request request, ResponseListener listener) {
```

Comment:
This took me a considerable amount of time to figure out what was actually happening in this method. Could you add some comments to help future readers determine what this method is doing?